[work in progress] Coroutines the Third

Aidan63 opened the pull request:
Third times the charm, right?

This is an evolution of #11554, the core state machine transformation has remained pretty much untouched from that branch, but the surrounding scaffolding code has been changed to more closely align with kotlin's coroutine implementation. This should hopefully address the issues surrounding continuation, scheduling, and a few others which were discovered in the other PR.

Changes

Coroutine functions are now transformed to accept a `Continuation<Any>` argument, return `Any`, and contain the state machine transformation. The continuation argument either holds the state of a paused execution of the coroutine function, or what will be invoked once the coroutine completes. The any return will either be the result of the coroutines execution, or a special token indicating that it's execution has been suspended.

```haxe
@:coroutine function foo(i:Int):String {}

function foo(i:Int, _hx_continuation:ICoroutine<Any>):Any {}
```

In addition each coroutine function has a class generated for it. This class implements `IContinuation<Any>` and stores data related to the execution of a coroutine, this includes, the current state, the result or error, variables accessed across states, and a reference to the continuation which will be invoked once this coroutine completes, essentially creating a chain. The continuation interface contains a `resume` function which when called resumes execution of the coroutine and is responsible to scheduling the execution of the coroutine into an appropriate thread.

```haxe
class HxCoro__Main_foo implements IContinuation<Any> {
	public var _hx_result : Any;

	public var _hx_error : Exception;

	public var _hx_state : Int;

	public var _hx_context : CoroutineContext;

	public var _hx_completion : IContinuation<Any>;

	public function resume(result:Any, error:Exception) {
		_hx_result = result;
		_hx_error  = error;

		_hx_context.scheduler.schedule(() -> {
			try {
				var result = Main.foo(0, this);
				if (result is Primitive) {
					return;
				} else {
					_hx_completion.resume(result, null);
				}
			} catch (exn:Exception) {
				_hx_completion.resume(null, exn);
			}
		});
	}
}
```

Coroutines can be started with the blocking function `Coroutine.run`, which blocks until a coroutine completes. Internally it has its own event loop which all coroutine continuations are scheduled onto.

A quick hello world example is provided below. It should work on all sys targets, but some targets might need their generators updated.

```haxe
import haxe.coro.Coroutine;
import haxe.coro.Coroutine.delay;

@:coroutine function test() {
    fancyPrint('hello');

    delay(1000);

    fancyPrint('world');
}

@:coroutine function fancyPrint(text:String):Void {
    for (i in 0...text.length) {
        Sys.print(text.charAt(i));

        delay(100);
    }

    Sys.print('\n');
}

function main() {
    Coroutine.run(test);
}
```

Resources

I'm putting a few links below, specifically related to kotlin's coroutines, which helped the puzzle come together a bit for me.

https://kt.academy/article/cc-under-the-hood

That is a good overview of the underlying machinery of kotlin's coroutines. Jumping straight into decompiled kotlin is very daunting as there are many, many layers of abstraction due to optimisations, features, etc. That article cuts through all of it and gets right to the core of whats happening. I pretty much copied that for this coroutine attempt.

https://www.droidcon.com/2022/09/22/design-of-kotlin-coroutines/

This then goes one layer deper, it covers a lot of those layers you see in kotlin's implementation and explains what they're there for.

https://github.com/JetBrains/kotlin/blob/master/compiler/backend/src/org/jetbrains/kotlin/codegen/coroutines/coroutines-codegen.md

Finally, the mega document. Goes in depth into the jvm codegen of kotlin's coroutines.

Simn commented:
I noticed that we were stuck on an outdated hxcpp branch. After updating this we now have green CI for cpp, which is a good baseline.

As the next step I'd like to get the JVM tests green. And afterwards we'll have to figure out how to make all this work on non-sys targets, because our coroutine execution currently relies on `sys.thread.EventLoop`.

Simn commented:
Update:

* *cpp* still works
* *jvm* works as well now after [fixing a crazy inference problem](https://github.com/Aidan63/haxe/commit/d4431e49c42db0a03f6b476f2338d1cde32cefc6)
* On *js*, the coro tests are passing too after [Aidan implemented a non-sys EventLoop](https://github.com/Aidan63/haxe/commit/03c264cf3d0245960a7676032472eb6fe92d4b79). CI is still red here because the server tests will need an updated version of what @klabz did in https://github.com/kLabz/utest/commit/1b9e193e9516af3cd3579ecbbbd4509fa0a7f534
* *eval* runs into #12176
* *neko* and *python* do not run coro tests yet
* *flash*, *php* and *lua* run into https://github.com/Aidan63/haxe/issues/7, and also don't run the coro tests yet
* *hl* has two failing tests:
```
TestBasic
  testResumeWithError: FAILURE F
    line: 30, exception of type haxe.Exception not raised
  testSimple: FAILURE F
    line: 6, expected 42 but it is 330584
```

Maybe @yuxiaomao could take a look at that some time. We also don't seem to run these tests on CI yet because HL is green.

Simn committed:
Deal with sys (#28)

* try something

* use the silly EntryPoint approach for RacingContinuation

* check the right flag

* check the right flag at the right place

Simn committed:
split up BaseContinuation

Simn committed:
bring back ContinuationControl

this doesn't do anything yet

Simn committed:
fix name

Aidan63 committed:
Remove now un-needed nested functions for the local function case

closes #12

Simn committed:
start injecting things

this still doesn't do anything

Simn committed:
Merge branch 'kt_coro' into kt_coro_tag

Simn committed:
it's mostly broken but some tests pass

Simn committed:
do what Aidan says

see #31

Simn committed:
remove debug print

Simn committed:
try different approach to typing

Simn committed:
turn getOrThrow into a coro

Simn committed:
getOrThrow shouldn't be a coro

Simn committed:
get something working with some hacks

Simn committed:
avoid hxcpp problem

Simn committed:
update HL

Simn committed:
redesign RacingContinuation

Simn committed:
disable failing tests for now

see #4

Simn committed:
disable misc test that fails for silly reasons

Simn committed:
Merge branch 'kt_coro' into kt_coro_tag

Simn committed:
Merge branch 'development' into kt_coro

# Conflicts:
#	src/context/typecore.ml
#	src/typing/typeloadCheck.ml
#	src/typing/typeloadFields.ml
#	src/typing/typer.ml

Simn committed:
clean up a bit

Simn committed:
Merge branch 'kt_coro' into kt_coro_tag

Simn committed:
let's see if we can run HL

Aidan63 committed:
expand coroutine function types for proper forward declarations

Aidan63 committed:
Merge pull request #33 from Aidan63/kt_coro_tag

_hx_control

Simn commented:
Update:

* *cpp*, *jvm* and *js* still work.
* *hl* works now too after reworking our data design in https://github.com/Aidan63/haxe/pull/33
* *eval* still fails until someone looks at #12176
* https://github.com/Aidan63/haxe/issues/7 has been resolved and so the CI for the affected targets is green now.
* *neko*, *python*, *php* and *lua* don't run the coro tests yet and probably all have some problem to sort out.
* *flash* will not be updated to support coroutines unless somebody steps up to do so.

Simn committed:
add @:coroutine.transformed (#39)

Simn committed:
add delay test

closes #34

Simn committed:
get neko working

Simn committed:
make test output file consistent

Simn committed:
apparently python just works

Simn committed:
Run tests on php and lua (#43)

* try php

* and lua

Simn committed:
Merge branch 'development' into kt_coro

Simn committed:
Merge branch 'development' into kt_coro

Simn committed:
remove invalid extra type parameter

Simn committed:
Make coros exception-free (#40)

* make coros exception-free

* avoid double looping if there's no catch

* add more exception tests

* add test for nested try-catch

* try to get exce[topm value management right

* port remaining YieldTryCatch cases

* make nested test even crazier

* change throw to break to avoid inner exceptions

* fix indentation

* also test throw from child coro

* ignore some tests on lua because there's a codegen issue

* lua isn't ready for this

* try again

* Move recursion tests back to flow control suite

* Fix missing import

---------

Co-authored-by: Aidan Lee <aidan.lee63@gmail.com>

Simn committed:
Add type parameters to BaseContinuation and ContinuationResult (#45)

* make coros exception-free

* avoid double looping if there's no catch

* add more exception tests

* add test for nested try-catch

* try to get exce[topm value management right

* port remaining YieldTryCatch cases

* make nested test even crazier

* change throw to break to avoid inner exceptions

* fix indentation

* also test throw from child coro

* ignore some tests on lua because there's a codegen issue

* lua isn't ready for this

* try again

* it works I guess

* also add to BlockingContinuation to lose the casts

* run JVM tests through hxb roundtrip

* be more accurate

* deal with LocalFunc type paramters too

* use more precise type on suspend

* T for all

Simn commented:
Easter update:

* *cpp*, *jvm*, *js* and *hl* still work
* *neko*, *python*, *php* and *lua* work too now after fixing minor issues for each of them (except python which just worked)
* *eval* still fails until someone looks at #12176
* We have refined our data design again in https://github.com/Aidan63/haxe/pull/40 by making coroutines exception-free. Users can still work with exceptions normally, but internally there are no exceptions being thrown around. Our hope is that this will make various aspects easier to reason about down the line.

Aidan63 committed:
Copy js' custom event loop, but add mutexes

Aidan63 committed:
Merge pull request #47 from Aidan63/custom-eval-loop

Get eval test suite running

Simn committed:
cache continuation_api on typer globals

Simn committed:
remove some API types that we don't need

Simn committed:
pass texprs as record type

The reason I prefer this is because it prevents us from accidentally messing up the order.

Simn committed:
clean up some outdated comments

Simn committed:
remove some unused internal data

Simn committed:
Optimize CFG a bit (#48)

* optimize

* another flag

* add missing catch handling, enable reindexing

* remove inner recursion from texpr transformer

* move functions to where they should be

Simn committed:
implement TCO (#49)

Simn committed:
add _hx_tmp local to deal with intermediate values (#53)

Simn committed:
don't forward blocks if their cb_catch differs

Simn committed:
rework error resume, again

closes #54

Simn committed:
actually we don't need the cb_catch check now, I think

Simn committed:
Add -D coroutine.throw for eager throws (#52)

* implement TCO

* add -D coroutine.throw for eager throws

* adjust

Aidan63 committed:
Keep waiting if we haven't got a result but the loop reports no more events.

Simn committed:
set next state later

closes #55

Simn committed:
use Unit instead of Void

closes #56

Simn committed:
change how we run JVM tests on CI

Simn committed:
try a new way of expression building

Simn committed:
move local handling to own function

this should probably actually move to coro.ml because the only thing it needs from here are the states

Simn committed:
Merge branch 'development' into kt_coro

Simn committed:
rename some things

Simn committed:
update resume-flag when forwarding blocks

closes #59

Simn committed:
Remove unreachable block (#60)

* remove cb_unreachable, use options instead

* add -D coroutine.noopt

* test all combinations

* nobody saw that

Simn committed:
pull structural changes from call-stacks branch

Simn committed:
Merge branch 'development' into kt_coro

# Conflicts:
#	src/generators/genjvm.ml

Simn committed:
pull null_pos changes from call-stacks branch

Aidan63 committed:
always use the mangled name with the variable id

Aidan63 committed:
no more _hx_ prefixes for core classes / interfaces

Aidan63 committed:
Merge pull request #62 from Aidan63/prefixes-and-collisions

Prefixes and collisions

Simn committed:
Merge branch 'development' into kt_coro

Simn committed:
Merge branch 'development' into kt_coro

yuxiaomao committed:
[std/hl] move some position for customStack and set_stack (#12217)

Aidan63 committed:
Call stacks (#50)

* Update stack item just before suspension

* class field and local func stack items

* wrap thrown exceptions in a CoroutineException

* bung the exception stack in the result field

* remove commented out code

* move buildCallStack to end of function

* add takeExceptionCallStack

* unbreak CI

* fix some types

* don't let null_pos mess up JVM codegen positions

* change while to if

* ignore for now

broken by the CoroutineException wrapping

* get it working on JVM

* how could I forget about Flash

* coroStack can be null

* make it work on eval

* apparently that can raise

* move top stack handling to BlockingContinuation

this way it works for immediate exceptions too

* remove _hx_stackItem, work with _hx_result directly

* remove CoroutineException, modify error.stack directly

* move texpr builder to context

* embrace the builder

* even more builder

* update __exceptionStack when setting the stack

this should allow reentrancy with regards to exceptions

* add something resembling a test

* overdesign test case

* disable hxb roundtrip for now because that must be an unrelated haxe problem

* this won't work in throw-mode

* make call stack tests print more info on failures

* add __customStack assignment to HL

* remove null_pos from our lives

* cpp setting custom stack

* Add forgotten actual stack setting

* hack around top stack problem a bit to test cpp and eval

* Revert "disable hxb roundtrip for now because that must be an unrelated haxe problem"

This reverts commit d78590d4ad34f16f1a584fcac4033041443334a6.

---------

Co-authored-by: Simon Krajewski <simon@haxe.org>

Simn committed:
send all types to analyzer for purity inference

see #12224

RblSb committed:
Test for #12224 (#12227)

RblSb committed:
Test for #11909 (#12218)

RblSb committed:
[nullsafety] Allow statics init in main (#12211)

* [nullsafety] Allow statics init in main

* usage before init check

RblSb committed:
Update TestStrict.hx (#12228)

kLabz committed:
4.3.7 changelog

Aidan63 committed:
interfaces seemingly working

Aidan63 committed:
Add test

Aidan63 committed:
number issue properly

Aidan63 committed:
Add an assert.pass

we just want to know if this even compiles

Aidan63 committed:
Fix coroutine overload resolution in jvm

Aidan63 committed:
Expand coroutine arguments for php

Aidan63 committed:
Merge pull request #70 from Aidan63/interfaces

Interfaces

Simn committed:
Merge branch 'development' into kt_coro

Simn committed:
get call stack test passing on HL

Simn committed:
fix test

Simn committed:
rename state to gotoLabel and control to state (#73)

Simn committed:
never mind

PXshadow committed:
haxe/coro: add Mutex typedef for threaded targets, and stub for non (#72)

* haxe/coro: add Mutex typedef for threaded targets, and stub for non threaded targets

* haxe/coro: add body to Mutex stub constructor

* tests: add TestMutex

* more publicity

* fix package

* try to make doc gen happy

---------

Co-authored-by: Simon Krajewski <simon@haxe.org>

Simn committed:
don't try to optimize resume states

closes #75

Simn committed:
avoid null issue

see #75

Simn committed:
also avoid JVM issue

Simn committed:
use ctx.type_params for type params

closes #77

Simn committed:
Merge branch 'development' into kt_coro

Simn committed:
Extensible context (#78)

* get started on extensible contexts (again)

* introduce Element after all

* more API

* use BalancedTree instead of Array

Simn committed:
assign value blocks to temp var

closes #79

Simn committed:
change Element to IElement (#80)

Simn committed:
be more diligent with positions

Simn committed:
fix more optimization problems

Simn committed:
[WIP] Stack traces continued (#82)

* improve stack traces again

* rework it again

Simn committed:
Try a new stack trace approach (#84)

* good start potentially

* more hacks required

* remove wonky sync stack handling

* don't try to manage stack traces on JS because we can't

* avoid cpp test problem

Simn committed:
Merge branch 'development' into kt_coro

Simn committed:
Make Context immutable (#87)

* make context immutable

* actually make it immutable

Simn committed:
remove -D coroutine.throw (#89)

Simn committed:
disable optimizations for now

too many distractions, will revisit this later

Simn committed:
deal with branching expressions in value places

closes #90

Simn committed:
Merge branch 'development' into kt_coro

Simn committed:
Merge branch 'development' into kt_coro

Simn committed:
assign suspension calls in value places to local

closes #93

Simn committed:
Revert "assign suspension calls in value places to local"

This reverts commit cc67f42d32557a72c32cda9f339a71cb6ad2945a.

Simn committed:
awkwardly deal with stack suspend problem

Aidan63 committed:
Don't use the typers current module for manged class names

Aidan63 committed:
Merge pull request #95 from Aidan63/cont_name_clash

Fix continuation name clash

Aidan63 committed:
per class anon function counter

Aidan63 committed:
Merge pull request #96 from Aidan63/stable_anon_naming

Per class anon function counter

Simn committed:
fix transformer bug and add more tests

Simn committed:
change context data structure to linked list (#100)

Simn committed:
Structured Concurrency with Task/Scope (#97)

* try yet another approach

* factor out startChildren

* Explicit type hint to make cpp generator happy

* adjust hierarchy once more

* bring back scope.with

I didn't notice the test was missing

* change entry point handling to support Coroutine.with

* add more tests and align with structured branch some more

* rework cancellation

* add wasResumed instead

* remove join, move loop handling back to entrypoint

* remove completion callbacks, instead store continuations directly

* adjust to feedback

* Add catch segv define to see if we get a stack trace in the future

* organize and document a bit

* refactor some more

* move scope addChild to its constructor

* support Coroutine.with().create

That's a neat looking API!

* scheduler experiments ported to new structured branch

* fix wasResume handling

* don't get distracted by optimization test insanity

* start children when beginning completion

This is much faster, but we should go back to counting completed children to make it faster still.

* fix wasResume handling

* work on event loop linked list from both ends

* go back to counting children to determine completion

* child scope test class uses virtual time scheduler

* optimize ms == 0 cases

* Other tests involving delay go through the virtual time scheduler

* make CI green because I want to see if PHP still explodes

* Try a double buffer for zero events

* switch schedule arguments around

* Make virtual time scheduler work how you'd expect it to work

* rename Scope to Node

* Move functions out of haxe.coro.Coroutine (#102)

* move coro API to hxcoro.Coro

* missed one

* move coro runner API to hxcoro.CoroRun

* missed another one

* change context to property

closes #101

* remove double allocation from suspend

closes #103

---------

Co-authored-by: Aidan Lee <aidan.lee63@gmail.com>
Co-authored-by: Aidan Lee <Aidan63@users.noreply.github.com>

Simn committed:
unset children when they completed

closes #105

Simn committed:
add some micro benchmarks

Aidan63 committed:
Add some virtual time scheduler tests

Aidan63 committed:
scheduler functions can be cancelled

Simn committed:
fix complex wrapper expressions

Aidan63 committed:
Callback on abstract task for cancellation

Aidan63 committed:
handle for stopping cancellation callback

Aidan63 committed:
Cancellation token approach

Simn committed:
add Context.add

Simn committed:
walk event loop from both ends

Aidan63 committed:
Merge branch 'kt_coro' into prompt-cancellation

Aidan63 committed:
Cancellation token key

Aidan63 committed:
static no-op cancellation handle

Aidan63 committed:
lazy initialise cancellation callback and children array

Aidan63 committed:
lets get fancy after all

Simn committed:
Add hxcoro.ds.Channel (#107)

* add hxcoro.ds.Channel

* maybe that's too much for some targets

* remove mutex

* fix wonky logic

* fix wonky logic differently and test with random delays

* change underlying data structure to paged deque

* move PagedDeque to its own file

Simn committed:
clean up a little and document random things

Aidan63 committed:
document new interfaces

Aidan63 committed:
Merge pull request #110 from Aidan63/prompt-cancellation

Prompt cancellation

Aidan63 committed:
timeout implementation

Aidan63 committed:
Add some tests

Simn commented:
Since I'm accidentally opening PRs from the coroutine branch already anyway, I thought I should give an update on the state of affairs!

## Lower level

The lower level, which is the transformation of `@:coroutine` marked functions to state machines that support suspension, has been pretty stable for a few weeks now. We still occasionally come across bugs in the transformer, but the overall architecture is quite solid and works well on all targets.

There currently isn't much optimization (by which I mean it is outright disabled) because we want to focus on stability first. But even without that, the performance we're seeing is quite good.

## Structured concurrency

We've been looking into the higher level in the form of "coroutine tasks", with a strong focus on structured concurrency. This is best explained as a code example:

```haxe
import haxe.Timer;
import hxcoro.CoroRun;
import hxcoro.Coro.*;

function main() {
	final startTime = Timer.stamp();
	final result = CoroRun.runScoped(node -> {
		final infiniteTicker = node.async(_ -> {
			var i = 0;
			while (true) {
				trace(i++ * 200);
				delay(200);
			}
		});

		final lazyTask = node.lazy(_ -> trace("We're done"));

		final slowTask = node.async(node -> {
			trace("slowly doing something");
			var aggregateValue = 0;
			var completedChildren = 0;
			for (_ in 0...500) {
				node.async(_ -> {
					while (Math.random() > 0.01) {
						yield();
						aggregateValue++;
					}
					completedChildren++;
				});
			}
			delay(1500);
			trace('$completedChildren children completed');
			return aggregateValue;
		});

		final result = slowTask.await();
		infiniteTicker.cancel();
		lazyTask.start();
		return result;
	});
	trace('done after ${Timer.stamp() - startTime} seconds, aggregate value is $result');
}
```

This creates a task tree which looks like this:

```
top
    infiniteTicker
    lazyTask
    slowTask
        500 child tasks
```

And it behaves like this:

![cmd_0QvzRyLI9W](https://github.com/user-attachments/assets/899398bb-6d0c-4983-9cc0-14bcf447f332)

Even though there's no actual threading yet (there will be), structured concurrency allows us to run multiple tasks "at the same time", which we can see by the overall run-time of the program being just slightly above the 1500ms from the top task's delay. The example also shows that we support cooperative cancellation (`infiniteTicker.cancel()`) and lazy tasks (`lazyTask.start()`).

## Data structures

Something else we've started working on are coroutine-aware data structures. For now we only have `hxcoro.ds.Channel`, which is a queue that allows asynchronous reading and writing like this:

```haxe
import hxcoro.ds.Channel;
import haxe.Timer;
import hxcoro.CoroRun;
import hxcoro.Coro.*;

function main() {
	final startTime = Timer.stamp();
	var aggregateValue = 0;
	CoroRun.runScoped(node -> {
		final channel = new Channel();
		final numTasks = 10000;
		// set up writers
		for (_ in 0...numTasks) {
			node.async(_ -> {
				delay(Std.random(1000));
				channel.write(1);
			});
		}
		// set up readers
		for (_ in 0...numTasks) {
			node.async(_ -> {
				delay(Std.random(1000));
				aggregateValue += channel.read();
			});
		}
	});
	trace('done after ${Timer.stamp() - startTime} seconds, aggregate value is $aggregateValue');
}
```

Compiling this to JavaScript, we get an outcome like this:

```
source/Main.hx:27: done after 1.1163641000166535 seconds, aggregate value is 10000
```

The result value being 10000 confirms that all writers and readers did indeed run, and the overall execution time shows that we have low overhead (we can expect at least one task to delay to the full 1000ms, so that's the baseline). This also scales quite well, though how far we can take this still depends largely on the target:

numTasks | hl | cpp | js | jvm
--- | --- | --- | --- | ---
1000 | 1.062 | 1.003 | 1.022 | 1.026
10000 | 1.589 | 1.087 | 1.116 | 1.037
100000 | 8.618 | 1.878 | 1.638 | 1.266
1000000 | 78.568 | crash | oom | 4.045

Looks like there's more to optimize here yet!

## hxcoro

We have decided to keep the lower level in Haxe's std, but move everything that's not strictly related to it into a hxcoro haxelib. For the time being, the sources for this haxelib are still in std itself because that simplifies development a lot, but this will eventually be moved outside.

Simn committed:
separate (#117)

Simn committed:
generate continuation stack traces only in debug mode

Simn committed:
avoid more array allocations

Aidan63 committed:
Extra test

Aidan63 committed:
Document timeout

Aidan63 committed:
Merge branch 'kt_coro' into timeout

Aidan63 committed:
Merge pull request #116 from Aidan63/timeout

Timeout

Aidan63 committed:
Bring back the event loop mutexes

Simn committed:
implement awaitChildren (#123)

Aidan63 committed:
release future lock before calling the function

Aidan63 committed:
Merge remote-tracking branch 'origin/kt_coro' into kt_coro

Simn committed:
don't hold onto zeroMutex while we're running

Simn committed:
also remove double loop because run itself is run in a loop

Simn committed:
Don't reschedule BaseContinuation.resume (#118)

* don't reschedule BaseContinuation.resume

* go back to scheduling somewhat

Simn committed:
disable linux lua because this is annoying

Simn committed:
Merge branch 'development' into kt_coro

Simn committed:
even less lua

Simn committed:
go away...

Simn committed:
move task stuff to hxcoro.task

Simn committed:
add some API to PagedDeque

Simn committed:
add produce sample as a test because it happens to work right now

see #124

Simn committed:
Don't have RacingContinuation hold onto its mutex for so long (#127)

* don't have RacingContinuation hold onto its mutex for so long

also avoid some unnecessary scheduling

* go for yet another approach

* inline by hand because HL hates me

* avoid capturing `this`

Simn committed:
Add CoroMutex (#112)

* start over without task ID

* move to hxcoro.concurrent

* change API to match haxe.atomic.AtomicInt

* move to own file

* use native implementations where possible

* fixes and tests

Simn committed:
make task context lazy and avoid some retention

Simn committed:
add CoroTask.putOnHold and call it from CoroSemaphore

Simn committed:
make some more context retention adjustments

Simn committed:
enable optimizations again to see if we have broken anything in the meantime

Simn committed:
ignoring failing eval call stack test

Simn committed:
fix some potential null problems

Aidan63 committed:
Cancelling suspend (#129)

* first pass at cancelling suspend

* tests

* Close handle on resuming

Otherwise callback could be invoked later on

* bounce continuation call through the scheduler

* Work with ICancellationHandle instead of its closure (#131)

* work with ICancellationHandle directly instead of closures

* rename to ICancellationCallback.onCancellation

* adjust to changes awkwardly

* add mutex cancellation test

* is it really that easy?

* add more elaborate semaphore cancelling test

* Immediate invoke callback if already cancelled

* Disallow multiple assignment of the callback

* flip the words

* change to cancellable

* I think this atomic state makes sense

* Interface name change for consistency

* document

* let's play sure it's not about this...

* less handle terminology

---------

Co-authored-by: Simon Krajewski <simon@haxe.org>

Simn committed:
Merge branch 'development' into kt_coro

Simn committed:
Merge branch 'development' into kt_coro

Aidan63 committed:
Prompt channels (#126)

* prompt write cancellation

* Use AssertAsync class I added a while back

* FIFO suspended writes

* prompt read cancellation

* actually assert the promptness of cancellations

also expected is actually actual

* go back to PagedDeque

* remove unintended double resume

curious that this doesn't cause anything to fail

* adapt

* Failing junction test (#128)

* add funny test that fails on some targets

* adapt

* fix order of operations

* change producer example to infinite loops

* use static extensions and port prime example

* change to bufferSize

---------

Co-authored-by: Simon Krajewski <simon@haxe.org>

Simn committed:
finally add an id to tasks

Simn committed:
optimize common CancellationHandle.close case

Simn committed:
make Key interface nicer

Simn committed:
don't use @:native

Simn committed:
add hxcoro.components.CoroName

closes #67

Simn committed:
get the old generator tests working

see #5

Simn committed:
add Coro.supervisor

Simn committed:
move parent back to AbstractTask

Simn committed:
add C type parameter to tasks (#135)

Simn committed:
Add INodeStrategy (#136)

* add INodeStrategy, favor composition over inheritance

* use concrete parameters after all

* Revert "use concrete parameters after all"

This reverts commit 2e7ff6849cac200199ce5a0eeca418937aec540e.

* revert type param changes

* split up CoroTask

Simn committed:
Add hxcoro.util.Convenience (#134)

* add hxcoro.Convenience

* adapt

* adapt more

Simn committed:
Merge branch 'development' into kt_coro

Simn committed:
refine some toString

Simn committed:
Merge branch 'development' into kt_coro

Simn committed:
fix double suspension_result wrapping

see #139

Simn committed:
fix some field hosts

see #139

Aidan63 committed:
Schedulers use Int64 ms everywhere

Aidan63 committed:
Change float literal

Aidan63 committed:
Int64 literal to make some targets happy

Aidan63 committed:
i64 literal here as well, for paranoia's sake

Aidan63 committed:
Private access Int64.toString to make some targets happy

Simn committed:
spell things out for HL

Simn committed:
spell it out for every other target too

Simn committed:
Add ILocalContext (#137)

* add ILocalContext

* use for awaitingChildContinuation

Aidan63 committed:
Merge pull request #140 from Aidan63/milliseconds

Millisecond Schedulers

Simn committed:
add initialState to AbstractTask constructor and `start` automatically if Running

closes #124

Simn committed:
[mad science] Get rid of explicit .key expressions (#141)

* add Key.fromClass to get rid of explicit .key expressions

* we actually don't care about the class itself at all

Simn committed:
add Channel benchmark

Simn committed:
remove expression mapping that doesn't seem to be needed anymore

Simn committed:
bring back some mapping that was needed after all

Aidan63 committed:
re-enabled virtual time scheduler tests

Aidan63 committed:
Implement yield as delay 0

Simn committed:
don't ignore RLocal when mapping complex expressions

closes #146

Aidan63 committed:
Remove private access int64 workaround

Simn committed:
Clean up _hx_tmp handling (#145)

* split up hx_tmp for result/error

* try to get error wrapping under control

* add surprisingly simple tests

Simn committed:
clean up result handling a little

Simn committed:
make most Context functions inline

Simn committed:
add missing return

Simn committed:
Add IScheduleObject (#143)

* add IScheduleObject

* use for continuations

* use for async

* also use for CancellationContinuation

* implement suspendCancellable directly

closes #138

* avoid some double-scheduling

* use AtomicInt instead of simon-mutexes for RacingContinuation too

* factor out delayImpl so both delay and yield can call it

This avoids the double suspension call from yield to delay, which I noticed in the yield benchmarks.

Simn committed:
go back to Array context because that's simply faster

Simn committed:
reduce BaseContinuation data a little more

Simn committed:
Merge branch 'development' into kt_coro

# Conflicts:
#	src/typing/typerEntry.ml

Simn committed:
Merge branch 'development' into kt_coro

Simn committed:
eval has atomics now

Simn committed:
Change isCancellationRequested to cancellationException (#147)

* change isCancellationRequested to cancellationException

* bring back isCancellationRequested as a convenience method

Simn committed:
Make internal coroutine initialization lazy (#151)

* start moving some types to continuation api

* more

* more

* don't load exception twice

* be lazier

* be even lazier

* fix

* ignore that

Simn committed:
Merge branch 'development' into kt_coro

Simn committed:
Add PagedDeque.remove (#154)

* support element deletion in Page

* get cpp green even though this seems weird

* add some more tests

* fix isEmpty but find more problems

* fix isEmpty resetting

* don't blit with length 0

* use ArraySort for stability

* change API to something that might actually be sound

* test middle page deletion

* move tests to TestPagedDeque

Simn committed:
Merge branch 'development' into kt_coro

Aidan63 committed:
Merge branch 'HaxeFoundation:development' into kt_coro

Aidan63 committed:
Follow away abstracts before getting the default value

Aidan63 committed:
Add test

Aidan63 committed:
Add an assert

Aidan63 committed:
Merge pull request #161 from Aidan63/follow_for_default

Follow for default

Simn committed:
treat {} as the special case that it is

Simn committed:
Merge branch 'development' into kt_coro

# Conflicts:
#	README.md

Simn committed:
haxelib

kLabz committed:
[tests] fix server tests (#164)

* [tests] fix server tests builder

* WIP fix server tests with coro

* [CI] enable server tests..

* Use promise wrapper instead of blocking run

* point to my updated utest coro branch

* Fix server display tests

---------

Co-authored-by: Aidan Lee <aidan.lee63@gmail.com>

Simn committed:
Merge branch 'development' into kt_coro

Simn committed:
support static extensions

closes #167

Aidan63 committed:
Merge branch 'HaxeFoundation:development' into kt_coro

Aidan63 committed:
Merge branch 'HaxeFoundation:development' into kt_coro

Aidan63 committed:
Add default value int64 test

Aidan63 committed:
Fix scheduled events not being ordered by time

Aidan63 committed:
Manual Int64 equality check

Aidan63 committed:
Spilling (#152)

* Some initial work on variable spilling

* skip restoring with the first state

* bodge saving and restoring

* make arguments appear in the first state usage table

* Don't restore variables in their defined state

* Add loop iteration hoisting test

* another attempt

* sort out arguments

they're always a pain

* round and round we go...

Back to per state vars, was hoping the "tmp used without being initialised" would magically solve itself, but no. Might be related to TVar exprs

* give some vars default type expressions

* bodge it

* disable hanging test so we can look at actual failures

* hack it even more

* Add some comments so I don't forget what this all is again

* attempt at avoiding double wrapping

seems to work on some targets...

* mark them as captured as well

* don't duplicate half of capturedVars

* Fix dodgy merge

* Need to follow abstracts away before getting default values

Not sure why I need to do this now

---------

Co-authored-by: Simon Krajewski <simon@haxe.org>

Aidan63 committed:
Merge pull request #168 from Aidan63/kt_coro_event_times

EventLoopScheduler times

Aidan63 committed:
Spilling Reads and Writes (#170)

* track local read and writes for states

* Use Texpr.skip

Simn committed:
Channel API (#150)

* add tryRead

* progress so far

* work around variance problem

* apply variance to bounded reader as well

* Add reader wakeup tests

* add write tests

* adjust

* generic createUnbounded

* load of bounded reader tests

* Remove test which I think we don't need

* channel kind enum

* add reader tests to main

* work around weird paged dequeue issue

* channel tests updated and passing

* bring back all tests

* Add basic creation tests

* add writer closing tests

* remove some unused imports

* reader cancellation

* async success, but immediate scheduler in tests

* Fix missing var

* Move existing tests into the bounded test class

* Writer sets out closed object

* Add another tests which includes the wait for read and channel closing

* use PagedDeque.remove

* go back to the dequeue tryPop loop

* add concurrent circular buffer

* configurable write behaviour when buffer is full

* experiments with single reader and writer combo

* drop callbacks

* Quickly thrown together and untested ubbounded channel

* prompt cancellation waitForWrite

* throw channel closed when writing to a closed channel

* Add unbounded writer tests

* work around weird issue

* disable single producer consumer implementation

will revisit at a later time

* Work around weird python issue by breaking these out into variables

* unbounded reader tests

* Add tryPeek to channel reader

* Have bounded channel use a circular buffer instead of an array

---------

Co-authored-by: Aidan Lee <aidan.lee63@gmail.com>

Simn committed:
Merge branch 'development' into kt_coro

Aidan63 committed:
Add thread pool scheduler (#159)

* Add thread pool scheduler

* Use event loop now

* Make negative time exception message consistent

* try sys wrap

* fix thread dodge

---------

Co-authored-by: Simon Krajewski <simon@haxe.org>

kLabz committed:
[std] move hxcoro package to hxcoro

And add basic 3rd party tests (eval only for now)

kLabz committed:
[tests] remove tests moved to hxcoro

Aidan63 committed:
Move continuations to be the leading argument

Aidan63 committed:
Use the hxcoro branch

Aidan63 committed:
Merge pull request #174 from Aidan63/leading_continuations

Move continuations to be the leading argument

Aidan63 committed:
Merge remote-tracking branch 'origin/kt_coro' into kt_coro

barisyild commented:
Is there any plan for https://github.com/HaxeFoundation/haxe/issues/12406 issue? 

This is a big limitation for creating production-ready http servers.

barisyild commented:
Is a ThreadLocal-like implementation possible for coroutines?

https://github.com/hxwell/hxwell/blob/f0723379b43a9d97c25088fbbb50ca07aba640ed/src/haxe/ThreadLocal.hx

Simn commented:
For deployment reasons, the `kt_coro` branch now is on the HaxeFoundation remote: https://github.com/HaxeFoundation/haxe/tree/kt_coro

I'll go ahead and close this PR accordingly. More updates hopefully soon...
